azavea/ashlar-blueprint
# Add simple frontend

# Overview

Adds a simple frontend to the reference implementation allowing public users to view and filter Records.

### Checklist


### Demo

View all Records for a given Record Type on the map:

![frontend](https://user-images.githubusercontent.com/14170650/41994097-630067a6-7a1b-11e8-8e16-0d395582f5b2.png)

Or, add a filter:

![frontend-filtered](https://user-images.githubusercontent.com/14170650/41994109-6da5294e-7a1b-11e8-90c2-b8115ed847a4.png)

### Notes

- Most of the JS is written inline in the main template, [`frontend/app/templates/index.html`](https://github.com/azavea/ashlar-blueprint/blob/57e0e9917e20eb5c80f9f647f56c8acf0bcf3292/frontend/app/templates/index.html#L83). This isn't a pattern that I stand behind necessarily, but it seemed like the simplest and clearest way to do it with vanilla JS. If you have opinions about how it could be modularized while still maintaining the clarity, I'm happy to hear it!
- Some of the logic for retrieving/sending filters is quite a doozy; in particular [`getJsonbFilter`](https://github.com/azavea/ashlar-blueprint/blob/57e0e9917e20eb5c80f9f647f56c8acf0bcf3292/frontend/app/templates/index.html#L176) and  [`updateFilters`](https://github.com/azavea/ashlar-blueprint/blob/57e0e9917e20eb5c80f9f647f56c8acf0bcf3292/frontend/app/templates/index.html#L265), both of which have to store data in deeply-nested objects and run complicated switches to determine the format of the filters/queries. I'm also not confident I got all of the filtering logic right for all possible filters, since there are quite a few possible combinations. The amount of time I had to spend getting my head around it seems to me to be further evidence that generating queries/filters could be a great place to start for an [external JS library](https://github.com/azavea/ashlar-2018-fellowship/issues/20).

## Testing instructions

- Make sure your containers are up to date and start a dev server:

```
./scripts/clean && ./scripts/update && ./scripts/server
```

- Poke around, try filters, and see what works! To find values for filtering, you can click on markers to see a JSON blob with their data. You can also log into the schema editor at `localhost:9000` and to make changes and try different fields.
